pkg/rbd/rbd_util.go

```diff
@@ -0,0 +1,225 @@
+	secret := b.adminSecret
+	if id == "" {
+		id = b.userID
+		secret = b.userSecret
+	}
+
+	glog.V(4).Infof("rbd: status %s using mon %s, pool %s id %s key %s", image, b.monitors, b.pool, id, secret)
+	args := []string{"status", image, "--pool", b.pool, "-m", b.monitors, "--id", id, "--key=" + secret}
+	cmd, err = execCommand("rbd", args)
+	output = string(cmd)
+
+	if err, ok := err.(*exec.Error); ok {
+		if err.Err == exec.ErrNotFound {
+			glog.Errorf("rbd cmd not found")
+			// fail fast if command not found
+			return false, output, err
+		}
+	}
+
+	// If command never succeed, returns its last error.
+	if err != nil {
+		return false, output, err
+	}
+
+	if strings.Contains(output, imageWatcherStr) {
+		glog.V(4).Infof("rbd: watchers on %s: %s", image, output)
+		return true, output, nil
+	} else {
+		glog.Warningf("rbd: no watchers on %s", image)
+		return false, output, nil
+	}
+}
+
+// DeleteImage deletes a ceph image with provision and volume options.
+func DeleteRBDImage(image string, pOpts *rbdVolumeOptions) error {
+	var output []byte
+	found, _, err := rbdStatus(image, pOpts)
+	if err != nil {
+		return err
+	}
+	if found {
+		glog.Info("rbd is still being used ", image)
+		return fmt.Errorf("rbd %s is still being used", image)
+	}
+	// rbd rm
+	l := len(pOpts.monitors)
+	// pick a mon randomly
+	start := rand.Int() % l
+	// iterate all monitors until rm succeeds.
+	for i := start; i < start+l; i++ {
+		mon := pOpts.monitors
+		glog.V(4).Infof("rbd: rm %s using mon %s, pool %s id %s key %s", image, mon, pOpts.pool, pOpts.adminID, pOpts.adminSecret)
+		args := []string{"rm", image, "--pool", pOpts.pool, "--id", pOpts.adminID, "-m", mon, "--key=" + pOpts.adminSecret}
+		output, err = execCommand("rbd", args)
+		if err == nil {
+			return nil
+		}
+		glog.Errorf("failed to delete rbd image: %v, command output: %s", err, string(output))
+	}
+	return err
+}
+
+func execCommand(command string, args []string) ([]byte, error) {
+	cmd := exec.Command(command, args...)
+	return cmd.CombinedOutput()
+}
+
+func getRBDVolumeOptions(volOptions map[string]string, client *kubernetes.Clientset) (*rbdVolumeOptions, error) {
+	rbdVolume := &rbdVolumeOptions{}
+	var ok bool
+	var err error
+	rbdVolume.adminID, ok = volOptions["adminId"]
+	if !ok {
+		return nil, fmt.Errorf("Missing required parameter adminId")
+	}
+	adminSecretName, ok := volOptions["adminSecretName"]
+	if !ok {
+		return nil, fmt.Errorf("Missing required parameter adminSecretName")
+	}
+	rbdVolume.adminSecretNamespace, ok = volOptions["adminSecretNamespace"]
+	if !ok {
+		rbdVolume.adminSecretNamespace = "default"
+	}
+	rbdVolume.adminSecret, err = parseStorageClassSecret(adminSecretName, rbdVolume.adminSecretNamespace, client)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve Admin secret %v", err)
+	}
+	rbdVolume.pool, ok = volOptions["pool"]
+	if !ok {
+		return nil, fmt.Errorf("Missing required parameter pool")
+	}
+	rbdVolume.monitors, ok = volOptions["monitors"]
```

Comment:
The challenge with this is that the Monitor endpoints could change. So this Volume information could get staled, which will lead to a failure when the image is being mapped. Can the monitor endpoints be obtained dynamically? In Rook, we have a configmap that gets updated whenever the monitor endpoints change (ie a new monitor gets added etc).

Reply:
csi provides monitor information to the plugin and it takes it from the storage class which is served by this specific plugin.
Please keep in mind that it is first csi rbd plugin implementation, where we try to offer the same level of functionality as currently offered by the existing in kubernetes rbd driver. At this point there is no goal to match what Rooks offers.

Reply:
@sbezverk but even in the case outside of rook. If a new monitor is deployed and joins the quorum, is the expectation that the admin would have to create a new storageClass to include this monitor?

Reply:
To answer your question, you do not need to create a new storage class, you can edit the existing one and add new monitor there.

Reply:
ref https://github.com/kubernetes-incubator/external-storage/pull/537 from @dmick

Reply:
I will keep an eye on this PR, thanks for the link.